addons/purchase/models/account_invoice.py

```diff
@@ -177,11 +177,37 @@ def _anglo_saxon_purchase_move_lines(self, i_line, res):
                             for val_stock_move in valuation_stock_move:
                                 valuation_price_unit_total += abs(val_stock_move.price_unit) * val_stock_move.product_qty
                                 valuation_total_qty += val_stock_move.product_qty
+
+                            # in Stock Move price unit is in company_currency
                             valuation_price_unit = valuation_price_unit_total / valuation_total_qty
                             valuation_price_unit = i_line.product_id.uom_id._compute_price(valuation_price_unit, i_line.uom_id)
+                        else:
+                            valuation_price_unit = i_line.purchase_line_id.product_uom._compute_price(i_line.purchase_line_id.price_unit, i_line.uom_id)
+                            # The PO, hence the price unit, can be in another currency
+                            # convert it to company currency
+                            order_id = i_line.purchase_line_id.order_id
+                            if order_id.currency_id != order_id.company_id.currency_id:
+                                valuation_price_unit = (
+                                    order_id.currency_id
+                                    .with_context(date=i_line.purchase_line_id.order_id.date_approve)
+                                    .compute(valuation_price_unit, order_id.company_id.currency_id, round=False)
+                                )
+
+                    line_price_unit_company_currency = i_line.price_unit
                     if inv.currency_id.id != company_currency.id:
-                            valuation_price_unit = company_currency.with_context(date=inv.date_invoice).compute(valuation_price_unit, inv.currency_id, round=False)
-                    if valuation_price_unit != i_line.price_unit and line['price_unit'] == i_line.price_unit and acc:
+                        line_price_unit_company_currency = (
+                            inv.currency_id
+                            .with_context(date=inv.date_invoice)
+                            .compute(line_price_unit_company_currency, company_currency, round=False)
+                        )
+
+                    # Valuation price unit and line_price_unit_company_currency in Company Currency
+                    # line['price_unit'] and i_line.price_unit both in the currency of the invoice, foreign or not
```

Comment:
I don't think we can make this assumption, although it's going to be true in most cases.